File: diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int solve(TreeNode *root,int &ans){
        if(root==NULL)
            return 0;
        int left_path=solve(root->left,ans);
        int right_path=solve(root->right,ans);
        ans=max(ans,left_path+right_path);
        return max(left_path,right_path)+1;
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans=0;
        solve(root,ans);
        return ans;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def solve(node):
            nonlocal ans
            if node is None:
                return 0 
            left = 0 
            right = 0
            left = solve(node.left)
            right = solve(node.right)
            # if this node have maxium diameeter then 
            ans = max(ans,left+right)
            return max(left,right)+1
        if root is None:
            return 0
        solve(root)
        return ans